loaders/dashboards/pkg/controller/dashboard_controller.go

```diff
@@ -259,17 +270,51 @@ func updateDashboard(old, new interface{}, overwrite bool) {
 		}
 	}
 
-	for _, value := range new.(*corev1.ConfigMap).Data {
-
+	for key, value := range new.(*corev1.ConfigMap).Data {
 		dashboard := map[string]interface{}{}
 		err := json.Unmarshal([]byte(value), &dashboard)
 		if err != nil {
-			klog.Error("Failed to unmarshall data", "error", err)
+			klog.Error("Failed to unmarshal data", "error", err)
 			return
 		}
-		if dashboard["uid"] == nil {
-			dashboard["uid"], _ = util.GenerateUID(new.(*corev1.ConfigMap).GetName(),
-				new.(*corev1.ConfigMap).GetNamespace())
+
+		name := new.(*corev1.ConfigMap).GetName()
+		namespace := new.(*corev1.ConfigMap).GetNamespace()
+
+		// Check to see if the dashboard contains a uid.
+		if dashboard["uid"] == nil || dashboard["uid"] == "" || dashboard["uid"] == "null" {
+			klog.Infof("no uid detected for dashboard %v. generating new uid for dashboard", name)
+			dashboard["uid"], _ = util.GenerateUID(namespace, name)
+
+			data, _ := json.Marshal(dashboard)
+			new.(*corev1.ConfigMap).Data[key] = string(data)
+
+			_, err := coreClient.ConfigMaps(namespace).Update(context.TODO(), new.(*corev1.ConfigMap),
+				metav1.UpdateOptions{})
+			if err != nil {
+				klog.Warningf("failed to update configmap: %v", err)
+			}
+		}
+
+		uid := dashboard["uid"].(string)
+
+		if _, ok := uidentifierList[uid]; !ok {
+			uidentifierList[uid] = name
+		} else if uidentifierList[uid] != name {
+			klog.Warningf("Dashboard named '%v' has the same uid. generating new uid for dashboard '%v'",
```

Comment:
what's the best option if duplicate uid detected?
to generate new uid, or reject the import for the dashboard?

Reply:
https://github.com/stolostron/multicluster-observability-operator/blob/bea1ed3ef1e448e1478af0ee1a10175a089aa595/operators/multiclusterobservability/pkg/util/clustermanagementaddon.go#L24

cluster overview dashboard's id is used here. We need additional check to guarantee the id not changed for this dashboard.